io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -728,27 +746,26 @@ object IO {
    * Submerges the error case of a disjunction into the `IO`. The inverse
    * operation of `IO.attempt`.
    */
-  final def absolve[E, A](v: IO[E, E \/ A]): IO[E, A] =
-    v.flatMap {
-      case -\/(e) => IO.fail(e)
-      case \/-(a) => IO.now(a)
-    }
+  final def absolve[E, A, B](disj: B => Disj[E, A])(v: IO[E, B]): IO[E, A] =
+    v.flatMap(b => disj(b).fold[IO[E, A]](IO.fail, IO.now))
+
+  @inline
+  private[effect] def absolveEither[E, A](sinner: IO[E, Either[E, A]]) =
+    IO.absolve[E, A, Either[E, A]](eitherDisj)(sinner)
 
   /**
    * Retrieves the supervisor associated with the fiber running the action
    * returned by this method.
    */
-  def supervisor[E]: IO[E, Throwable => IO[Void, Unit]] = new Supervisor()
+  def supervisor[E]: IO[E, Throwable => IO[Nothing, Unit]] = new Supervisor()
 
   /**
    * Requires that the given `IO[E, Maybe[A]]` contain a value. If there is no
    * value, then the specified error will be raised.
    */
-  final def require[E, A](error: E): IO[E, Maybe[A]] => IO[E, A] =
-    (io: IO[E, Maybe[A]]) => io.flatMap(Maybe.maybe(IO.fail[E, A](error))(IO.now[E, A]))
+  final def require[E, A, B](error: E): IO[E, Disj[Unit, A]] => IO[E, A] =
```

Comment:
After looking at all this, it does make me wonder if we should just use `Either` / `Option`, then rely on the opaque types in Scalaz 8 to ensure we use `Either` / `Option` to back `\/` / `Maybe`.

Reply:
I'm interested in this idea, I felt the new API sort of unergonomic after using it for a wile here internally and in the tests. I'm more that open to change it to something better. You mean, port (copypasta) just the needed bit of `\/` and `Maybe` from sz8 to here? Renaming them to avoid ambiguities I reckon.

Reply:
No, I mean use `Either` and `Option` instead of `Disj`. We don't need it "on the way in", because the API is already elegant for things like `attempt`, where you specify the success and failure cases. But "on the way out", the use of `Disj` does make the code feel a bit nasty. I think maybe it's better to delete `Disj` and just use `Either` and `Option` in those places. We can take care in Scalaz 8 core to use opaque types so we can do cost-free versions for `\/` and `Maybe`.